frontend/simple/js/transactions.js

```diff
@@ -0,0 +1,263 @@
+'use strict'
+
+import localforage from 'localforage'
+import * as Events from '../../../shared/events'
+import sbp from '../../../shared/sbp'
+
+let _steps = new WeakMap()
+let _scope = new WeakMap()
+let _cursor = new WeakMap()
+let _description = new WeakMap()
+let _isRecoverable = new WeakMap()
+let _wait = new WeakMap()
+// Why would need this?  to make our retries grow in a common sense ways
+function fibSequence (num) {
+  let output = 0
+  let sequence = [ 0, 1 ]
+  for (var i = 0; i < num; i++) {
+    sequence.push(sequence[ 0 ] + sequence[ 1 ])
+    sequence.splice(0, 1)
+    output = sequence[ 1 ]
+  }
+  return output
+}
+
+let transactionDB = localforage.createInstance({ name: 'Transactions', storeName: 'Transactions' })
+export class Transaction {
+  constructor (description, recoverable) {
+    _description.set(this, description)
+    _steps.set(this, [])
+    _scope.set(this, {})
+    _isRecoverable.set(this, !!recoverable)
+    _cursor.set(this, 0)
```

Comment:
This `WeakMap` stuff seems like a very strange way of doing things... going to try to figure out why it's done like this, and if there's another way to do it

Reply:
Its just private membership based on instances instead of scope. Its not that complicated but wasn't possible before es6. A weakmap will keep a reference to the mapped data until the instance at its key is garbage collected.